### PR TITLE
Hardening systemd unit for additional security

### PR DESCRIPTION
**Pull request by paulfantom**
Add a couple of options to restrict what prometheus process can see and do.

Mostly taken from:
- https://github.com/konstruktoid/hardening/blob/master/systemd.adoc#unit-configuration
- https://prog2017.rmll.info/IMG/pdf/2017-07-05_rmll2017_durcissement_avec_systemd.pdf
- https://gist.github.com/ageis/f5595e59b1cddb1513d1b425a323db04
- https://www.freedesktop.org/software/systemd/man/systemd.exec.html

PR needs some extensive testing to ensure everything works.

### CODE REVIEW

**Comment by SuperQ** [comment]
Why add the extra config option?  All of these changes look sane enough that users shouldn't need to disable it.

**Comment by paulfantom** [comment]
I worry that on older systems it might cause problems. That's why I added an extra config option which by default is set to `True`.

For example `ProtectSystem=strict` is available since systemd 232.

**Comment by SuperQ** [comment]
Ahh, yes, it would be good to keep compatibility with at least Debian/Jessie and Ubuntu/16.04.  I think that specific feature may need to stay out of the default since 16.04 only has 229.

**Comment by paulfantom** [comment]
I did some research:

| OS | Systemd version |
| ------------- | --- |
| Debian Jessie | 215 |
| Centos 7 | 219 |
| Ubuntu 16.04 | 229 |
| Debian Stretch | 232 |
| Ubuntu 18.04 | 237 |
| Fedora 27 | 238 |

Feature used in PR with related systemd version (first occurance):

| Feature | Version | More info |
| ------- | ------- | --------- |
| `PrivateTmp` | pre 183 |  |
| `ProtectSystem` | 214 | `strict` option since 232 |
| `ProtectHome` | 214 | |
| `NoNewPrivileges` | ? | |
| `ReadWritePaths` | 231 | Previously `ReadWriteDirectories` since pre 199 |
| `ProtectControlGroups` | 232 | |

File: README.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ All variables which can be overridden are stored in [defaults/main.yml](defaults
 | `prometheus_version` | 2.2.1  | Prometheus package version. Also accepts `latest` as parameter. |
 | `prometheus_config_dir` | /etc/prometheus | Path to directory with prometheus configuration |
 | `prometheus_db_dir` | /var/lib/prometheus | Path to directory with prometheus database |
+| `prometheus_service_hardening` | True | Apply security related options to systemd service file |
```
Comment (SuperQ) [inline README.md]: Let's add a note about the minimum systmed version required for this.

File: defaults/main.yml [hunk 2]
```diff
@@ -4,6 +4,8 @@ prometheus_version: 2.2.1
 prometheus_config_dir: /etc/prometheus
 prometheus_db_dir: /var/lib/prometheus
 
+prometheus_service_hardening: True
```
Comment (SuperQ) [inline defaults/main.yml]: I think we should disable this by default, as the older operating systems are still quite popular.

**Comment by paulfantom** [comment]
Fun fact, systemd ignores options it doesn't understand.

I ran this PR against debian jessie (systemd 215) with all hardening options enabled and prometheus works. Journalctl displays some warnings and an error:

> May 16 08:16:18 debian-s-1vcpu-1gb-fra1-01 systemd[1]: [/etc/systemd/system/prometheus.service:18] Failed to parse protect system value, ignoring: strict
May 16 08:16:18 debian-s-1vcpu-1gb-fra1-01 systemd[1]: [/etc/systemd/system/prometheus.service:19] Unknown lvalue 'ProtectControlGroups' in section 'Service'
May 16 08:16:18 debian-s-1vcpu-1gb-fra1-01 systemd[1]: [/etc/systemd/system/prometheus.service:20] Unknown lvalue 'ProtectKernelModules' in section 'Service'
May 16 08:16:18 debian-s-1vcpu-1gb-fra1-01 systemd[1]: [/etc/systemd/system/prometheus.service:21] Unknown lvalue 'ProtectKernelTunables' in section 'Service'

"Unknown lvalue" are just ignored, and error can be fixed by lowering security a little and applying `ProtectSystem=full` instead of `ProtectSystem=strict`.

> May 16 08:16:20 debian-s-1vcpu-1gb-fra1-01 prometheus[1379]: level=info ts=2018-05-16T08:16:20.667936858Z caller=main.go:588 msg="Loading configuration file" filename=
May 16 08:18:01 debian-s-1vcpu-1gb-fra1-01 systemd[1]: [/etc/systemd/system/prometheus.service:19] Unknown lvalue 'ProtectControlGroups' in section 'Service'
May 16 08:18:01 debian-s-1vcpu-1gb-fra1-01 systemd[1]: [/etc/systemd/system/prometheus.service:20] Unknown lvalue 'ProtectKernelModules' in section 'Service'
May 16 08:18:01 debian-s-1vcpu-1gb-fra1-01 systemd[1]: [/etc/systemd/system/prometheus.service:21] Unknown lvalue 'ProtectKernelTunables' in section 'Service'


TL;DR; @SuperQ you were right, we shouldn't have a flag to disable security hardening :smile:

**Comment by SuperQ** [comment]
We could have a variable for `prometheus_systemd_protect_level: full`, so users can flip to strict if they want.

File: tasks/preflight.yml [hunk 3]
```diff
@@ -75,3 +75,10 @@
   with_items:
     - "{{ lookup('url', 'https://github.com/prometheus/prometheus/releases/download/v' + prometheus_version + '/sha256sums.txt', wantlist=True) | list }}"
   when: "('linux-' + (go_arch_map[ansible_architecture] | default(ansible_architecture)) + '.tar.gz') in item"
+
+- name: Get systemd version
+  shell: "systemctl --version | head -n1 | sed 's/systemd //g'"
```
Comment (SuperQ) [inline tasks/preflight.yml]: This might be less fragile:

```
systemd --version | awk '$1 == "systemd" {print $2}'
```

Comment (paulfantom) [reply tasks/preflight.yml]: I cannot use `systemd` command in test suite, since it is not detected in ubuntu and debian containers, but `systemctl` works fine and it is always installed with systemd. As for `awk` it is indeed better.

Comment (SuperQ) [reply tasks/preflight.yml]: Sorry, coffee is not kicked in yet.  That was supposed to be the same `systemctl` command.

Comment (paulfantom) [reply tasks/preflight.yml]: No problem :smile:

Ok, do you think there is anyting more to do in this PR?
